actioncable/lib/action_cable/connection/subscriptions.rb

```diff
@@ -63,8 +63,21 @@ def unsubscribe_from_all
       private
         delegate :logger, to: :connection
 
+        def normalize_identifier(identifier)
+          identifier = ActiveSupport::JSON.encode(identifier) if identifier.is_a?(Hash)
+          identifier
+        end
+
+        # If `data` is a Hash, this means that the original JSON
+        # sent by the client had no backslashes in it, and does
+        # not need to be decoded again.
+        def decode_hash(data)
+          data = ActiveSupport::JSON.decode(data) unless data.is_a?(Hash)
+          data.with_indifferent_access
+        end
+
         def find(data)
-          if subscription = subscriptions[data['identifier']]
+          if subscription = subscriptions[normalize_identifier(data['identifier'])]
             subscription
           else
             raise "Unable to find subscription with identifier: #{data['identifier']}"
```

Comment:
Doing this piecemeal within Subscriptions feels touchy. We're likely to miss spots, introduce bugs, and Subscription's job shouldn't be baby-sitting our message transfer encoding.

Would hoist this up to `AC::Connection::Base#receive(data)`. That should responsible for decoding our message envelope (decoding as JSON), its contents (decoding the identifier and data as Identity ("do nothing" decoding) or as JSON), then passing the decoded message along to `Subscriptions#execute_command` for dispatch.

Similarly, channel transmit and server broadcast should get out of the business of implementing message encoding.